EIPS/eip-2200.md

```diff
@@ -0,0 +1,318 @@
+  * Still doesn’t make the invariant explicit.
+  * Requires callers to supply more gas, just to have it refunded
+* Add contract metadata specifying per-contract EVM version, and only
+  apply SSTORE changes to contracts deployed with the new version.
+
+## Backwards Compatibility
+
+This EIP requires a hard fork to implement. No gas cost increase is
+anticipated, and many contracts will see gas reduction.
+
+Performing SSTORE has never been possible with less than 5000 gas, so
+it does not introduce incompatibility to the Ethereum mainnet. Gas
+estimation should account for this requirement.
+
+## Test Cases
+
+Below we provide 17 test cases. 15 of them covering consecutive two
+`SSTORE` operations are based on work [by
+@chfast](https://github.com/ethereum/tests/issues/483). Two additional
+case with three `SSTORE` operations is used to test the case when a
+slot is reset and then set again.
+
+| Code                               | Used Gas | Refund | Original | 1st | 2nd | 3rd |
+|------------------------------------|----------|--------|----------|-----|-----|-----|
+| `0x60006000556000600055`           | 412      | 0      | 0        | 0   | 0   |     |
+| `0x60006000556001600055`           | 20212    | 0      | 0        | 0   | 1   |     |
+| `0x60016000556000600055`           | 20212    | 19800  | 0        | 1   | 0   |     |
+| `0x60016000556002600055`           | 20212    | 0      | 0        | 1   | 2   |     |
+| `0x60016000556001600055`           | 20212    | 0      | 0        | 1   | 1   |     |
+| `0x60006000556000600055`           | 5212     | 15000  | 1        | 0   | 0   |     |
+| `0x60006000556001600055`           | 5212     | 4800   | 1        | 0   | 1   |     |
+| `0x60006000556002600055`           | 5212     | 0      | 1        | 0   | 2   |     |
+| `0x60026000556000600055`           | 5212     | 15000  | 1        | 2   | 0   |     |
+| `0x60026000556003600055`           | 5212     | 0      | 1        | 2   | 3   |     |
+| `0x60026000556001600055`           | 5212     | 4800   | 1        | 2   | 1   |     |
+| `0x60026000556002600055`           | 5212     | 0      | 1        | 2   | 2   |     |
+| `0x60016000556000600055`           | 5212     | 15000  | 1        | 1   | 0   |     |
+| `0x60016000556002600055`           | 5212     | 0      | 1        | 1   | 2   |     |
+| `0x60016000556001600055`           | 412      | 0      | 1        | 1   | 1   |     |
+| `0x600160005560006000556001600055` | 40218    | 19800  | 0        | 1   | 0   | 1   |
+| `0x600060005560016000556000600055` | 10218    | 19800  | 1        | 0   | 1   | 0   |
+
+## Implementation
+
+To be added.
+
+## Appendix: Proof
+
+Because the *storage slot's original value* is defined as the value
+when a reversion happens on the *current transaction*, it's easy to
+see that call frames won't interfere SSTORE gas calculation. So
+although the below proof is discussed without call frames, it applies
+to all situations with call frames. We will discuss the case
```

Comment:
To disambiguate, do I understand what you mean by these terms?

- "current transaction" mean the entire transaction initiated by an EOA
- "call frame" means an invocation of a contract

Reply:
Yes. Those are clear terms defined in EVM.